### ISSUE TITLE
Interleaving DeckGL layers with a base map

### ISSUE THREAD

**Issue by gallickgunner** [visgl/deck.gl#4603]
Hello, this is just a continuation #4529. After successfully integrating harp.gl's map with deck.gl, I was experimenting with things and found my Icon Layer laid on top of the buildings of the map.

Normally in mapbox this could be solved by using the `mapboxLayer` and adding the icon layer beneath the building layer but harp.gl doesn't provide that sort of functionality. 

In essence deckgl is just drawing to an overlay canvas which is laid on top of the base canvas (harpgl's map in this case) if I understood correct. Is there any way to inform the overlay canvas of the depth buffer of the base canvas so everything renders correctly?

An example screen. the red circles show where the icon should be behind the buildings.
![testt](https://user-images.githubusercontent.com/40718428/82376198-2ef86480-9a3b-11ea-94a8-fd6291f7d33b.png)

Thanks.

**Comment by Pessimistress** [comment]
To do that you need to have both deck and harp.gl render into the same WebGLContext. Does harp.gl support third-party renderers?

**Comment by gallickgunner** [comment]
Harp.gl is completely built upon Three.js afaik. I've looked into it's source and there definitely is Threejs running in the background. I have experience with three.js a bit and I know that you can construct a `WebGLRenderer` out of an existing context one. 

So do we provide the context created by Deck to harp or the one created by harp to deck? In both cases, what's the procedure for extracting or feeding in the context to deckgl?

Also harp.gl doesn't expose this functionality so I'll have to basically modify it's source, but if I can get this working, I'll open a PR on their site, maybe they'll accept it. Thanks.

**Comment by Pessimistress** [comment]
If you can get the WebGLRenderingContext out of harp.gl, then construct `Deck` with `gl` as a prop.

If harp.gl accepts an external WebGLRenderingContext, then you can construct `Deck` first and use `deckInstance.gl`.

Then the trick is that whenever deck or the base map is updated, both need to redraw. This can be done with (pseudocode):

```js
baseMap.on('render', () => deckInstance.redraw('basemap redrawn', true));

deckInstance.setProps({
  // Normally, when deck renders to the canvas, it will wipe the existing content and draw all layers.
  // This callback overrides the default behavior
  _customRender: redrawReason => {
    if (redrawReason === 'basemap redrawn') { 
     // base map just redrew, due to e.g. tiles being loaded
     // render all layers without clearing the canvas
     deckInstance._drawLayers(redrawReason, {clearCanvas: false}); 
   } else {
     // Triggered by other reasons - layer update, view state change, transition, etc.
     // Tell the base map to redraw first
     baseMap.render(); 
   } 
  }
})
```

You can see how it works in the arcgis module:

https://github.com/visgl/deck.gl/blob/3b19fa2dc35361c83da6dba2373a436dbed80f58/modules/arcgis/src/commons.js#L57-L62

And the mapbox module:

https://github.com/visgl/deck.gl/blob/96ebcf3065c232d3719e853be6ed8bbb06399518/modules/mapbox/src/deck-utils.js#L13-L20

**Comment by gallickgunner** [comment]
@Pessimistress  - Ok. I looked into it and managed to get most of the part done but there's a problem. I'm using React and when I looked into the source for the react component of the deck, there is already a custom render function [bound](https://github.com/visgl/deck.gl/blob/ba63a096cfeb65f0fb1a68ddd981cf99ba1700ee/modules/react/src/deckgl.js#L137). Any way to circumvent this?

**Comment by gallickgunner** [comment]
@Pessimistress - So according to Nicol [here](https://stackoverflow.com/questions/50014874/opengl-and-webgl-sampling-from-a-texture-attached-to-current-framebuffer) the feedback loop error arises even if you try to read from a texture attached to the framebuffer, meaning it's the active one. So it doesn't matter if you are writing to it currently, as long as it's active you'll get the error. Does this ring any bells for you? Just asking if you know anything similar coded in deckgl.

**Comment by Pessimistress** [comment]
The depth issue is likely caused by the basemap and deck.gl using different near/far planes. If you can figure out how harp.gl is calculating its near/far we can try match deck's viewport to that.

Looks like harp.gl is rendering into a FBO, and did not reset the context afterwards. You can force deck to render into the screen buffer by setting:

```js
deck.setProps({
  parameters: {framebuffer: null}
});
```

Alternatively, you can try the approach used by `@deck.gl/arcgis`, which renders into a custom FBO and then composite it with the basemap.

**Comment by gallickgunner** [comment]
@Pessimistress  - Sorry for the late response and I appreciate your help for such a specific case, thanks. I was actually testing this whole thing with a post-processing effect applied to the deck. When I removed it, the error regarding the feedback loop disappeared. As far as I remembered post-processing does introduce an intermediary framebuffer and maybe that's what's causing it. Anyways leaving this for now, I've info regarding the near/far setting.

Harpgl had default far plane [setting ](https://github.com/heremaps/harp.gl/blob/12c20680f18878e83a6c9ba3214f956de79a8acc/%40here/harp-mapview/lib/MapView.ts#L182)to a very high number. I tried changing it to `1000` as deck uses `1000` if I'm correct. But nothing happened. Next I found that harpgl has a number of [ways ](https://www.harp.gl/docs/master/doc/interfaces/harp_mapview.clipplanesevaluator.html)to calculate the clip planes. Some of these are `ElevationBased`, `Interpolated`, `TiltViewBased`, `Fixed`, etc. By default it uses the `TiltView` one. [Here's ](https://github.com/heremaps/harp.gl/blob/12c20680f18878e83a6c9ba3214f956de79a8acc/%40here/harp-mapview/lib/ClipPlanesEvaluator.ts#L646)the code for that, if that makes any sense. I've also tried trying all of these but it doesn't do anything to help solve the problem. Maybe we need to tweak or match the settings with deck.

**Comment by gallickgunner** [comment]
@Pessimistress - anything you found out from the above info?

**Comment by nzjony** [comment]
@gallickgunner, I'm a developer on harp.gl and awesome that you are trying to integrate with deck.gl, we tried a small prototype internally with deck.gl, but we didn't do anything as complicated as you mention.

In harp.gl, the near/far plane changes based on the tilt (at least when using the default `TiltView` as you mentioned), so I guess that would cause problems with the integration.

You could try to use `FixedClipPlanesEvaluator` and then fix it to some constant value which is synced with deck.gl, does that help?

**Comment by Pessimistress** [comment]
@nzjony Thanks for the insight!

@gallickgunner You can retrieve the near/far plane from deck like this:

```js
deck.setProps({viewState});
const viewport = deck.getViewports()[0];
const {near, far} = viewport.projectionProps;
```

**Comment by gallickgunner** [comment]
@nzjony - Hey, thanks for the support.

@Pessimistress - Ok, I'll see what I can do with this info and update here with any results I get. 

Thanks, you two.

**Comment by gallickgunner** [comment]
@Pessimistress  - Sorry for the delay, I was sick for a couple of days. Anyways bad news is, we are back to ground zero. 

I did a fresh code up of the above test scenario and I can't seem to get the intermediate result I posted earlier, no matter what I try. I've uploaded the working test case [here ](https://github.com/gallickgunner/Deckgl-Experiments/tree/harp%2Bdeck)in case you want to try running it yourself.

So there seems to be no change if I render both harp and deck into the same context. Also, I've tried passing harpgl the webgl context created by deck by modifying the source of `MapView` in harp and it throws errors. Note sure what causes these maybe @nzjony can give me some pointers?.

![123](https://user-images.githubusercontent.com/40718428/85948048-4e6c9200-b903-11ea-8067-09ca1d668043.png)
![234](https://user-images.githubusercontent.com/40718428/85948047-4ca2ce80-b903-11ea-81c7-feb86f0a1dff.png)

I'll keep trying and see if I can get those intermediate results back. Don't know how or what caused those. Sadly I didn't had the chance to commit that and had to install a fresh copy of OS :(

**Comment by nzjony** [comment]
@gallickgunner , thanks for mentioning the issue, which browser are you using?

I have an idea how to fix it, can you try https://github.com/heremaps/harp.gl/pull/1651 ?  It seems that on some browsers, we need to enable this to access the function `fwidth`

**Comment by gallickgunner** [comment]
> @gallickgunner , thanks for mentioning the issue, which browser are you using?
> 
> I have an idea how to fix it, can you try [heremaps/harp.gl#1651](https://github.com/heremaps/harp.gl/pull/1651) ? It seems that on some browsers, we need to enable this to access the function `fwidth`

I guess this is a browser issue. I tried enabling the extension and it says extension not supported. I'm using Chrome `Version 83.0.4103.116 (Official Build) (64-bit)`

![ext](https://user-images.githubusercontent.com/40718428/86001568-eb8eff80-b9c3-11ea-842c-d789922f9f02.png)

**Comment by nzjony** [comment]
@gallickgunner , what GPU are you using? It could be because your GPU doesn't support these functions?

**Comment by gallickgunner** [comment]
Sorry for the delay as I found some other useful info and sorry for not mentioning these details in the OP. My machine specs are
```
Windows 10,
Ryzen 3600
Nvidia GTX 1660
Chrome 83.0.4103.116 (Official Build) (64-bit)
```
Anyways I think I found what's causing this error. I just found out that when passing the context from harp's canvas to deckgl, I was asking for a webgl1 context. I corrected that by doing

`mapCanvas.getContext( 'webgl2', contextAttributes );`

And threejs throwed an error immediately. I was using three.sj v 0.114 and it was trying to create webgl1 context by default. 
![ctx](https://user-images.githubusercontent.com/40718428/86003800-2f373880-b9c7-11ea-92e4-7f92de69e44b.png)

After I updated to the latest 0.118 release which was released a few days ago, it fixed this by trying to create a webgl2 context by default. However now, I got the same shader error as above. So I guess the reason why I was getting the shader error above when passing deck's initialized context to harp was because deckgl was creating a webgl2 context which I was then passing to threejs through harp. On the contrary, since I wasn't passing a webgl2 context to deck when accessing the context through harp's canvas (`mapCanvas` in above code), I wasn't getting the error then.

So, in short the above shader error occurs due to passing webgl2 context. Using webgl1 removes the error.

Secondly, I tried updating harpgl as well to `0.17` I was using `0.14`. And now, I'm getting some depth results.
![1](https://user-images.githubusercontent.com/40718428/86012695-1f255600-b9d3-11ea-9347-2f80fc3cb808.png)

![2](https://user-images.githubusercontent.com/40718428/86012707-22204680-b9d3-11ea-8764-e2aba2e511d3.png)

![image](https://user-images.githubusercontent.com/40718428/86014132-cb1b7100-b9d4-11ea-80fe-70a8df639510.png)

![image](https://user-images.githubusercontent.com/40718428/86013963-99a2a580-b9d4-11ea-916f-520bd2e0c8e0.png)

So I think now I can try the clipping thing you guys mentioned. Will update with the results.

**Comment by nzjony** [comment]
@gallickgunner , thanks for the update, and glad you managed to figure it out with the different contexts.

We haven't yet upgraded to 0.118 internally, so maybe there is some issues we need to fix, I will try to do this soon. Can you try with 0.117 for now?

Ok, if deck's far plane is dynamic, then you will probably need to implement a new `ClipPlaneEvaluator` with the exact same formula for it to work. Otherwise you will always have some issues. Actually, feel free to create a PR so others who want to integrate with deck.gl can use the same `ClipPlaneEvaluator`.

**Comment by gallickgunner** [comment]
@nzjony - Hmm, do we really have to re-implement the whole formula? I mean we can just have a new `ClipPlaneEvaluator` where we just pass near/far from deck. That was the intention in using the `FixedPlanesEvaluator` but it makes the far plane fixed and don't know why but throws an error. We just need a way to pass the near/far on every viewState change I guess.

Also don't you think coordinate spaces might matter? Deckgl extracts all the frustum information in [common ](https://github.com/visgl/deck.gl/blob/54ee46fe28c81c874096d230d53725eebc8f058e/modules/core/src/utils/math-utils.js#L39 )space which according to the docs is a left-handed space with `y` pointing downward. I don't found any info on the space used in harpgl.

**Comment by nzjony** [comment]
@gallickgunner , small update, I upgraded harp.gl to support three 0.118, the problem is that that version creates a Web GL 2 context by default and we have some raw shaders which don't yet support web GL 2, we will fix these soon. As a work around, I create by default a `WebGL1Renderer` for all examples, this explains the problems you were having with `fwidth` as you mentioned above.

Yes, that should work fine, I didn't catch that you want to update it each frame, but of course that should work.

I don't think the coordinate spaces will matter because you are only interested in the depth values being comparable right? Unless I miss something?

I don't fully understand what `minNear` and `minFar` are doing, from reading the code it seems they ensure that the range of the near / far plane doesn't reduce beyond this limit, the defaults are probably ok in your case.

**Comment by gallickgunner** [comment]
Ok thanks for the info. I'll try creating a new one or maybe modify the fixedplane one by passing deck's dynamic far plane every frame. Let's see what happens.

**Comment by gallickgunner** [comment]
@nzjony - It seems something is broken in the clip planes evaluators. I've tried using the `Fixed` and `TopView` evaluators. The fixed one gives errors, The `TopView` one just shows the arcs with a base blue background without any map tiles.

I also tried passing values from deckgl to harp and it throws the same errors as `Fixed` evaluator. So I think there is a problem with passing hardcoded or fixed values. I'll try removing deckgl integration and see if the `Fixed` evaluator works on its own. If not, then there is definitely an issue with harp.

These are the errors I get using `Fixed` or passing values from deck.
![errs](https://user-images.githubusercontent.com/40718428/86444998-7de61a80-bcc6-11ea-9865-f41d31ab5f79.png)


Secondly, I tried using `InterpolatedClipPlanesEvaluator` and this fixes the depth problem on some angles. As far as I managed to understand this one, calculates near/far based on ground distances? I've yet to figure out what exactly does it do that solves the problem. The depth issue comes back again at varius tilt angles though. I guess this is a hint that we need to do modify the `Interpolated` evaluator to keep working with tilt changes. Any ideas or pointers would be helpful. 

Here's a gif to show the issue coming back at various angles.

![ezgif-3-7577356a3c55](https://user-images.githubusercontent.com/40718428/86447223-e08ce580-bcc9-11ea-8a20-843761b398b9.gif)

**Comment by Pessimistress** [comment]
@gallickgunner There are multiple things that could result in this. You may want to control harp.gl's FOV as well. Can you also try multiplying deck's `near` and `far` with the viewport's height?

**Comment by gallickgunner** [comment]
> @gallickgunner There are multiple things that could result in this. You may want to control harp.gl's FOV as well. Can you also try multiplying deck's `near` and `far` with the viewport's height?

FOV was already set when initializing the mapView. I'll try multiplying the near/far with the viewport height, but I doubt it's gonna work, seems something wrong with using hardcoded/fixed values.

**Comment by Pessimistress** [comment]
Are you updating the values on every render? Deck’s fat plane is dependent on pitch.

**Comment by gallickgunner** [comment]
> Are you updating the values on every render? Deck’s far plane is dependent on pitch.

Yeap. It's just, trying to set any value causes errors as mentioned above. Might be some bugs with harpgl.

**Comment by nzjony** [comment]
@gallickgunner , sorry for the late reply.

Setting the near / far plane for the `FixedClipPlanesEvaluator` should work fine.

The error messages you are getting is very strange. I can't see how changing the near / far plane would cause such an issue. Can you open up the error message and see what the stack trace is?

**Comment by gallickgunner** [comment]
Sorry for the late reply.

@Pessimistress  - multiplying by viewport didn't change anything, still getting those errors.

@nzjony - This is the stack trace for the error that appears when you scroll. If you don't do anything the initial error is the one about "No element array buffer....". That one has no stack trace.

![stack](https://user-images.githubusercontent.com/40718428/87225447-7d284500-c3a6-11ea-99f0-19701e803a80.png)

**Comment by nzjony** [comment]
@gallickgunner , that is a very strange issue.

I made a small test project using the `FixedPlaneEvaluator`, can you take a look: https://github.com/nzjony/harpglclipplanetest and see if you can somehow tweak it to recreate the problem

**Comment by gallickgunner** [comment]
> @gallickgunner , that is a very strange issue.
> 
> I made a small test project using the `FixedPlaneEvaluator`, can you take a look: https://github.com/nzjony/harpglclipplanetest and see if you can somehow tweak it to recreate the problem

Thanks. I'll look into it and see what's the trigger for causing this.

**Comment by gallickgunner** [comment]
@Pessimistress  @nzjony 

So I find out that the errors don't arise when you aren't using a shared context between deck and harp. When using a shared context the trigger for the error is passing any layer to deckgl. As long as deckgl doesn't draw any layer there are no errors. 

So I think the "No Element Array Buffer..." error is from deckgl. And this leads harpgl to throw the other webgl error regarding "UniformMat` when scrolling.

Secondly, regarding the clip planes, the `FixedClipPlanes` tends to show a blue plane obstructing the map view at different values usually at lower `far` values. The blue plane that shows up at random positions horizontally. It also changes its positions based on zoom/orientation.  For very small values `2-3` as deckgl's far plane value comes out to be, no matter how far you zoom in/out the blue plane doesn't go away.

![bp1](https://user-images.githubusercontent.com/40718428/87845157-69b53680-c8de-11ea-8d36-5218ce076d1c.png)

![bp2](https://user-images.githubusercontent.com/40718428/87845159-6c179080-c8de-11ea-897d-e757a5012bc2.png)

![bp3](https://user-images.githubusercontent.com/40718428/87845161-6d48bd80-c8de-11ea-9125-6dbe57fbe06f.png)

@nzjony  - The lines you coded in your project, 
```
// Edit the fixed clip plane evaluator at run time.
mapView.addEventListener(MapViewEventNames.Render, (_) => {
  cpe.farPlane *= 1.001;
});
```
tend to move this blue plane upwards like its animating. This only happens however if the plane is partial like in pic 1 or 2. If you provide a value such that the plane is like in pic 1, then it doesn't clear up (moves upwards) on its own.

**Comment by nzjony** [comment]
@gallickgunner , sorry for the confusion, that line is adjusting the `far` plane, I did that just to test that changing the `far` plane at run time works, and that it doesn't cause the issue you mentioned with the `No element array buffer....`. This is not to demonstrate any valid real world use case.

What units does deck.gl use for world space? In harp.gl, 1 world unit is equal to one meter, that is why, when you zoom out, you see the blue plane sooner, because each zoom level requires a bigger near / far plane range.

**Comment by gallickgunner** [comment]
> 
> What units does deck.gl use for world space? In harp.gl, 1 world unit is equal to one meter, that is why, when you zoom out, you see the blue plane sooner, because each zoom level requires a bigger near / far plane range.

As far as I know deck uses lng lat and meters as the default coordinate system. According to the docs

> deck.gl's common space plays the role of standard "world space", but there are a few important differences. The y axis is inverted, which means it's a left-handed coordinate system. The mercator "zoom" factor is applied as a common space transform (rather than in the view or projection transforms).

**Comment by nzjony** [comment]
@gallickgunner , ok thanks for the clarification.

Regarding the images above, are these generated from my demo? Or are they generated when passing the near / far plane values from deck.gl to harp.gl?

**Comment by gallickgunner** [comment]
> @gallickgunner , ok thanks for the clarification.
> 
> Regarding the images above, are these generated from my demo? Or are they generated when passing the near / far plane values from deck.gl to harp.gl?

Both, it's possible to generate these images from your demo as well. Much easier to see if you pass very small values as far plane (under 200).

**Comment by nzjony** [comment]
@gallickgunner, ok, strange. Yes, my demo does that my design, but with deck.gl, that is werid. Do you have some project I can take a look at to see how it is set?

**Comment by gallickgunner** [comment]
@nzjony - I don't have a functional one currently. But I can. It's the same thing as your demo though, the blue planes occur regardless of whether you pass values from deck to harp or pass hardcoded values like `3,10,100,...`

**Comment by nzjony** [comment]
@gallickgunner, most peculiar. I guess then the world units are different between harp.gl and deck.gl, but I guess we need to see a demo to check that out.

**Comment by gallickgunner** [comment]
> @gallickgunner, most peculiar. I guess then the world units are different between harp.gl and deck.gl, but I guess we need to see a demo to check that out.

Hmm, what exactly do you want to check out btw? If the blue planes are occurring without deck integrated then shouldn't we fix that first since this is only related to harp? I've lost track of what the actual problem is here. 

1) I still don't understand what these blue planes actually are and "why" do they show up.
2) If they are supposed to show up, why do they get stuck when we pass really small fixed values (2/3). As I mentioned earlier, no amount of zooming tends to remove these.

**Comment by nzjony** [comment]
@gallickgunner , the blue plane is simple where no geometry is rendered (just the background clear color is shown, with some fog).

I.e. if you have the far plane closer than the geometry, then the geometry won't be rendered. Check out: http://learnwebgl.brown37.net/08_projections/projections_perspective.html it also has some demo which shows playing around with the far plane.

Normally, the default clip plane evaluator in harp is `TiltViewClipPlanesEvaluator`, i.e. the near / far planes are calculated based on the height of the camera above the earth etc.

However in your case, you need to match the far plane with that of deck.gl, hence using the `FixedClipPlaneEvaluator`.

If you are seeing blue, then the far plane is too close and it should be extended.

Does that help?

**Comment by gallickgunner** [comment]
> @gallickgunner , the blue plane is simple where no geometry is rendered (just the background clear color is shown, with some fog).
> 
> I.e. if you have the far plane closer than the geometry, then the geometry won't be rendered. Check out: http://learnwebgl.brown37.net/08_projections/projections_perspective.html it also has some demo which shows playing around with the far plane.
> 
> Normally, the default clip plane evaluator in harp is `TiltViewClipPlanesEvaluator`, i.e. the near / far planes are calculated based on the height of the camera above the earth etc.
> 
> However in your case, you need to match the far plane with that of deck.gl, hence using the `FixedClipPlaneEvaluator`. Note, if you zoom out, but use the `FixedClipPlaneEvaluator`, then the far plane has to be updated, because there is more world that can be seen.
> 
> If you are seeing blue, then the far plane is too close and it should be extended.
> 
> Does that help?

Okay now I get it. Then this means, as you were asking earlier, that there is likely a unit mismatch between deck and harp. Since values like `2,3` are very small and show these planes in harp where as everything works fine in deckgl with the same values.

@Pessimistress  - your intake on this?

**Comment by Pessimistress** [comment]
deck.gl's geospatial "common space" is 512x512, so at {lng:0, lat:0} the earth's circumference (~4e7m) maps to 512 pixels. Because web mercator projection is non-linear, pixels/meter changes with the latitude of the viewport center. You can access this ratio via `viewport.metersPerPixel`.

When zooming in/out, the camera (projection matrix) does not change. The view matrix scales the model by `2**zoom`.

The far plane you are looking for is probably some combination of the following parameters:

```js
viewport.projectionProps.far * viewport.height / (2 ** viewport.zoom) / viewport.metersPerPixel
```

**Comment by nzjony** [comment]
@Pessimistress , do you know what the units are in world / camera space, i.e. before the projection matrix is applied?

**Comment by gallickgunner** [comment]
Sorry for the late response, was busy with another thing. 

@Pessimistress - So I tried messing around with those paramteres but couldn't manage to fully synchronize with harp.

@nzjony - I think we should ditch the `FixedPlaneEvaluator` and start working with the `Interpolated` one as this one's atleast giving results and we can improve this with trial and error. Here is a picture after tweaking some parameters.

![image](https://user-images.githubusercontent.com/40718428/88449384-f3757e80-ce5f-11ea-9608-b4034fcf7316.png)

The parameters to tweak here are definitely the ones related farplane i.e. `nearFarMultiplier` and `FarOffset`. The problem is that changing the orientation of the map (bearing/pitch/zoom) causes the far plane values to change automatically and the results are lost. So it's definitely the far plane synchronization issue.

We need to have a better understanding of how both deck and harp are calculating these far values in order to properly synchronize them. Hope we will get this going.

**Comment by nzjony** [comment]
@gallickgunner , it would be good to know why passing the values from deck.gl to the harp.gl's `FixedPlaneEvaluator` doesn't work. That would only be the case where the units in deck.gl are different, I read: https://deck.gl/docs/developer-guide/coordinate-systems and it says that the default units are meters, are you perhaps using a different coordinate system?

**Comment by gallickgunner** [comment]
> That would only be the case where the units in deck.gl are different, I read: https://deck.gl/docs/developer-guide/coordinate-systems and it says that the default units are meters, are you perhaps using a different coordinate system?

No I'm using the default.

>  @gallickgunner , it would be good to know why passing the values from deck.gl to the harp.gl's `FixedPlaneEvaluator` doesn't work.

True, what I actually meant when I said that we should ditch the `FixedPlanesEvaluator` was that we should try and come up with a formula or something so the `InterpolatedEvaluator` works at all angles and zoom levels since we can verify that. If we manage to achieve that then we can reverse engineer and find the missing factors that are required for passing values between deck/harp.

**Comment by nzjony** [comment]
@gallickgunner , do you have a demo application I can take a look at. 

I don't quite understand the need to reverse engineer, the far / near planes should be compatible between harp.gl / deck.gl in theory. I think I need to see a working demo to help any further unfortunately because otherwise I am just speculating.

**Comment by Pessimistress** [comment]
> do you know what the units are in world / camera space, i.e. before the projection matrix is applied?

@nzjony The unit is zoom-dependent, and equals to filling the whole Mercator world ([-180, -85], [180, 85]) with `512 * Math.pow(2, zoom)` units. I do not quite understand your statement that "in harp.gl, 1 world unit is equal to one meter", as Web Mercator projection is [nonlinear](https://en.wikipedia.org/wiki/Mercator_projection#/media/File:Mercator_with_Tissot's_Indicatrices_of_Distortion.svg).

I agree that it'll be the easiest to debug if @gallickgunner can share the working code.

**Comment by gallickgunner** [comment]
@nzjony  @Pessimistress 

Sorry for the delay, I was trying to setup a mock dataset as I wasn't allowed to push the original one. You can find the worknig demo [here](https://github.com/gallickgunner/Deckgl-Experiments/tree/harp%2Bdeck).

**Comment by gallickgunner** [comment]
Any issues in running the demo from the repo guys? Sorry was a little busy the past 1-2 weeks.

**Comment by nzjony** [comment]
@gallickgunner , sorry, I was away. If I get some time this week I will take a look

**Comment by nzjony** [comment]
@gallickgunner , I played a round with your demo a bit yesterday, but didn't get so far. I did notice however that the far plane stays constant when zooming out, see: https://github.com/nzjony/Deckgl-Experiments/commit/d6b6ed725d610943376117d91cdffc3e9b39f937 Why does the far plane not change when zooming out?

I also saw here: https://deck.gl/docs/api-reference/core/web-mercator-viewport that they recommend subclassing the `WebMercatorViewport` class so that it matches our internal web mercator projection.

I also looked into how mapbox does their world space and it seems (if I understood correctly) that they have their world coordinates between 0 and 360 degrees, i.e. it is just a shift from the longitude, see: https://github.com/mapbox/mapbox-gl-js/blob/main/src/geo/mercator_coordinate.js.

I think your best bet is to subclass and make sure that the functions to convert from mercator to world space match our projection class.

**Comment by gallickgunner** [comment]
@nzjony - 
Okay, I'll wait for @Pessimistress '  response to most of the questions.


> I think your best bet is to subclass and make sure that the functions to convert from mercator to world space match our projection class.

If I understood correct you mean to do a ditto conversion like in harp [here](https://github.com/heremaps/harp.gl/blob/12c20680f18878e83a6c9ba3214f956de79a8acc/%40here/harp-geoutils/lib/projection/MercatorProjection.ts#L72) inside deck by extending the `WebMercatorViewport`. 

If I'm correct again, then I think I need to change [these](https://github.com/visgl/deck.gl/blob/84844da9b2a968c106f849eae3b480b66cc44b57/modules/core/src/viewports/viewport.js#L173). Hoping to get some input from Pessimistress. I'll try doing that on my end in the meantime though.

**Comment by Pessimistress** [comment]
Hi, sorry for the late reply. I couldn't get the demo to run with the base map (`WorkerService: [worker-service-manager]: Invalid message initialized`).

**Comment by gallickgunner** [comment]
> `WorkerService: [worker-service-manager]: Invalid message initialized`)

Hey, I forgot about this issue too, got completely wrapped up with other projects. Don't know why the error occurs but you aren't seeing the map because I set it initially to show the problem that `FixedPlaneEvaluator` throws errors regarding webgl and uniform matrices. Just change [this ](https://github.com/nzjony/Deckgl-Experiments/blob/6b8bb4961981587432c42778f5b88f09aba0c375/src/Root.js#L91) line to any other evaluator (like the commented one below it) or increase the far plane to high values like `1000+` as mentioned in the comment.

**Comment by Pessimistress** [comment]
So you can't call `mapView.updateCamera()` because that gets overriden later during `render()`. You need to update the `clipPlanesEvaluator`:

```js
  getClipPlaneEvaluator(viewState) {
    const viewport = new WebMercatorViewport(viewState);
    const height = viewport.height;
    let {near, far} = viewport.projectionProps;
    near *= height * viewport.metersPerPixel;
    far *= height * viewport.metersPerPixel;
    return new FixedClipPlanesEvaluator(near, far - near);
  }

  onViewStateChange = ( { viewState } ) => {
    const coords = new GeoCoordinates( viewState.latitude, viewState.longitude );
    this.mapView.lookAt( coords, MapViewUtils.calculateDistanceFromZoomLevel( { focalLength: this.mapView.focalLength }, viewState.zoom + 1 ), viewState.pitch, viewState.bearing );
    this.mapView.zoomLevel = viewState.zoom + 1;
    this.mapView.clipPlanesEvaluator = this.getClipPlaneEvaluator(viewState);
  
    this.setState( { viewState } );
  }
```

I tried setting `clipPlanesEvaluator` during MapView's initialization. It doesn't crash but it doesn't render either. There might be a bug in harp?

**Comment by gallickgunner** [comment]
> So you can't call `mapView.updateCameras()` because that gets overriden later during `mapView.render()`. You need to update the `clipPlanesEvaluator`:
> 
> ```js
>   getClipPlaneEvaluator(viewState) {
>     const viewport = new WebMercatorViewport(viewState);
>     const height = viewport.height;
>     let {near, far} = viewport.projectionProps;
>     near *= height * viewport.metersPerPixel;
>     far *= height * viewport.metersPerPixel;
>     return new FixedClipPlanesEvaluator(near, far - near);
>   }
> 
>   onViewStateChange = ( { viewState } ) => {
>     const coords = new GeoCoordinates( viewState.latitude, viewState.longitude );
>     this.mapView.lookAt( coords, MapViewUtils.calculateDistanceFromZoomLevel( { focalLength: this.mapView.focalLength }, viewState.zoom + 1 ), viewState.pitch, viewState.bearing );
>     this.mapView.zoomLevel = viewState.zoom + 1;
>     this.mapView.clipPlanesEvaluator = this.getClipPlaneEvaluator(viewState);
>   
>     this.setState( { viewState } );
>   }
> ```
> 
> I tried setting `clipPlanesEvaluator` during MapView's initialization. It doesn't crash but it doesn't render either. There might be a bug in harp?

I tried your snippet and it's working on my end. Not getting any error or rendering issue. Tried logging the far value both on the deckgl side and inside `MapView.js` and it prints the same value too.

![image](https://user-images.githubusercontent.com/40718428/94119826-d9104f00-fe68-11ea-97c6-3bd353a4c715.png)

The only problem remaining is the original one, icons being blocked by buildings when they should be in front

**Comment by Pessimistress** [comment]
Your icons are anchored at the center instead of the base of the pin. You need to assign `anchorY` to the icon descriptor:

![image](https://user-images.githubusercontent.com/2059298/94353359-2f29f180-0025-11eb-9c7d-d63ee9c883ff.png)

You can push the far plane further away by specifying `farZMultiplier` in Deck's view state. The default is `1.01`.

**Comment by gallickgunner** [comment]
@Pessimistress  - Tried `achorY` with value `0` and equal to `height` but it doesn't seem to fully solve the issue.

![dept](https://user-images.githubusercontent.com/40718428/94357530-05b7a700-00b3-11eb-9605-ae72271248e6.png)

Depth checking goes haywire at the top of the icon when it's clearly in front but works for the bottom part of the icon. What's the actual reason for this? Is far plane mismatching the root cause for these type of issues?

Now I don't know how icons are working internally, but I'm guessing that when we are using them as always face to the camera, the 2D icons stand upright and coincide with the altitude axis. If they don't and are titled, then that will explain why some part of the icon go inside the building.  If the buildings and everything stand upright or if they are tilted then if everything (icons + buildings) is tilted at the same angle, I don't get why some things or parts go inside the buildings.

I know this sounds preposterous, cuz honestly I've never come across this scenario in my fair share of work with OpenGL etc, but is there any way to make it work so that the depth checking, checks for an anchor point to determine whether the icon is in front or back and then uses that sole point as a reference for the whole rendering.

**Comment by Pessimistress** [comment]
Yep you got it. The icon layer renders as billboards, i.e. all icons are drawn facing the camera at the depth of their anchor.

To draw them in real 3D you need each icon's rectangle in the world space instead of the clip space. You may switch to the SimpleMeshLayer with texture, or you can subclass the IconLayer and modify its `billboard: false` mode.

**Comment by gallickgunner** [comment]
> Yep you got it. The icon layer renders as billboards, i.e. all icons are drawn facing the camera at the depth of their anchor.
> 
> To draw them in real 3D you need each icon's rectangle in the world space instead of the clip space. You may switch to the SimpleMeshLayer with texture, or you can subclass the IconLayer and modify its `billboard: false` mode.
> 
> Edit: looks like the original issue (depth matching with harp.gl) is solved? We can move the icon rendering discussion to a different thread.

Hmm, I'll try that. And yes I think this was the main issue all along. So depth matching with harp.gl works already, I guess. There might be issues with far plane mismatching which give those yellowish/blue planes but more or less they can be avoided or tweaked.

I'll open a new thread regarding icon rendering when I'm back on this. Thanks for the continued support, appreciate it.

**Comment by nzjony** [comment]
@gallickgunner , congratulations! Glad to hear you got it working, and thanks @Pessimistress also for your help!
